plutus-core-spec/figures/PlutusCoreTypeCorrectness-Algorithmic-Unrestricted.tex

```diff
@@ -0,0 +1,232 @@
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\appT{A}{B}}{K}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{$tcn$ is a type constant in in Fig. \ref{fig:Plutus_core_type_constants}}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\sizeK{}}}\)}
+        \RightLabel{tycon}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istypeJ{\conT{tcn}{A}}{\typeK{}}}\)}
+    \end{prooftree}
+
+    \caption{Kind Synthesis}
+    \label{fig:Plutus_core_kind_synthesis}
+\end{figure*}
+
+
+
+
+
+\begin{figure*}[t]
+    \judgmentdef{\(\hypJ{\Gamma}{\istermJ{M}{S}}\)}{In context $\Gamma$, term $M$ has normal type $S$}
+
+    \begin{prooftree}
+        \AxiomC{\(\ctxni{\Gamma}{\termJ{x}{S}}\)}
+        \RightLabel{var}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{x}{S}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{$cn$ has constant signature $\constsig{tcn}{s}$ in Fig. \ref{fig:Plutus_core_constants}}
+        \RightLabel{con}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{cn}{\conT{tcn}{s}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma, \typeJ{\alpha}{K}}{\istermJ{M}{T}}\)}
+        \RightLabel{abs}
+        \UnaryInfC{\(\hypJ{\Gamma}{\istermJ{\abs{\alpha}{K}{M}}{\allT{\alpha}{K}{T}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istermJ{L}{\allT{\alpha}{K}{T}}}\)}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{K}}\)}
+        \AxiomC{\(\typeMultistep{\subst{A}{\alpha}{T}}{R}\)}
+        \RightLabel{inst}
+        \TrinaryInfC{\(\hypJ{\Gamma}{\istermJ{\inst{L}{A}}{R}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\Gamma \vdash Q :: \typeK{}\)}
+        \AxiomC{\(Q = \mathcal{E}\{\fixT{\alpha}{A}\}\)}
+		\AxiomC{\(\Gamma \vdash M : \mathcal{E}\{[\fixT{\alpha}{A}/\alpha]A\}\)}
+		\RightLabel{wrap}
+		\TrinaryInfC{\(\hypJ{\Gamma}{\istermJ{\wrap{\alpha}{A}{M}}{Q}}\)}
+    \end{prooftree}
+    
+    Where $\mathcal{E}$ is an elimination context $\mathcal{E} ::= \bullet \mid [\mathcal{E} \,\, A]$, $A$ a type, such that $\mathcal{E}\{Q\}$ denotes the type expression obtained by replacing the $\bullet$ in $\mathcal{E}$ with $Q$. That is, $[\bullet\,\, A]\{B\} = [B\,\,A]$, and so on. 
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istermJ{M}{\fixT{\alpha}{S}}}\)}
+        \AxiomC{\(\typeMultistep{\subst{\fixT{\alpha}{S}}{\alpha}{S}}{R}\)}
+        \RightLabel{unwrap}
+        \BinaryInfC{\(\hypJ{\Gamma}{\istermJ{\unwrap{M}}{R}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istypeJ{A}{\typeK{}}}\)}
+    	\AxiomC{\(\typeMultistep{A}{S}\)}
+        \AxiomC{\(\hypJ{\Gamma, \termJ{y}{S}}{\istermJ{M}{T}}\)}
+        \RightLabel{lam}
+        \TrinaryInfC{\(\hypJ{\Gamma}{\istermJ{\lam{y}{A}{M}}{\funT{S}{T}}}\)}
+    \end{prooftree}
+
+    \begin{prooftree}
+        \AxiomC{\(\hypJ{\Gamma}{\istermJ{L}{\funT{S}{T}}}\)}
+        \AxiomC{\(\hypJ{\Gamma}{\istermJ{M}{S'}}\)}
+        \AxiomC{\(S \equiv S'\)}
```

Comment:
Given that these are both normal types, I think this should be `=` not `equiv`? Generally I was expecting `equiv` not to appear in the algorithmic version.

Reply:
`≡` means "syntactically equal modulo alpha", whereas `=` means "`≡` modulo beta", so since we're being very explicit about where reduction happens, we should only ever be using `≡` in the type system, I think.

cc @mchakravarty who proposed the `=`/`≡` naming convention.

Reply:
(This comment was in the wrong place, sorry!)

Oh so there is - I got the symbols the wrong way around and I thought triple-equals was equality modulo beta. My notational instincts led me to assume assume that the triple-equals was "stronger" and hence was the one that included reduction.

Ugh, I should have another look through and make sure all the equalities are indeed the ones I expect, then.